### PR TITLE
feat(intersection): Full support of intersection types.

### LINKED ISSUE #471: add support for intersection type
see Section 4.9 of "The Java ® Language Specification Java SE 8 Edition", p 70
https://docs.oracle.com/javase/specs/jls/se8/jls8.pdf

### PR DESCRIPTION
**Pull request by GerardPaligot**
Closes #471

### CODE REVIEW

File: src/main/java/spoon/reflect/reference/CtTypeParameterReference.java [hunk 1]
```diff
@@ -27,31 +27,45 @@
 	/**
 	 * Gets the bounds (aka generics) of the referenced parameter.
 	 */
+	@Deprecated
 	List<CtTypeReference<?>> getBounds();
 
 	/**
-	 * Returns {@code true} if the bounds are upper bounds.
+	 * Returns {@code true} if the bounds are in <code>extends</code> clause.
```
Comment (monperrus) [inline src/main/java/spoon/reflect/reference/CtTypeParameterReference.java]: false means a "super clause"

File: src/main/java/spoon/support/reflect/declaration/CtTypeParameterImpl.java [hunk 2]
```diff
@@ -46,33 +44,44 @@ public void accept(CtVisitor v) {
 
 	@Override
 	public <T extends CtTypeParameter> T addBound(CtTypeReference<?> bound) {
-		if (bounds == CtElementImpl.<CtTypeReference<?>>emptyList()) {
-			bounds = new ArrayList<CtTypeReference<?>>(TYPE_BOUNDS_CONTAINER_DEFAULT_CAPACITY);
+		if (getSuperType() instanceof CtIntersectionTypeReference<?>) {
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/declaration/CtTypeParameterImpl.java]: else? in the fly creation of CtInterTypRef? or throw new Exception?

File: src/main/java/spoon/reflect/reference/CtTypeParameterReference.java (hunk 1 above)
Comment (GerardPaligot) [inline src/main/java/spoon/reflect/reference/CtTypeParameterReference.java]: Done.

File: src/test/java/spoon/test/type/TypeTest.java [hunk 3]
```diff
@@ -153,11 +160,63 @@ public void testIntersectionBindingReturnsFirstType() throws Exception {
...
+
+		final CtClass<Pozole> aPozole = launcher.getFactory().Class().get(Pozole.class);
+		final CtMethod<?> prepare = aPozole.getMethodsByName("finish").get(0);
+
+		// Intersection type in generic types.
+		final List<CtClass> localTypes = prepare.getElements(new TypeFilter<>(CtClass.class));
+		assertEquals(1, localTypes.size());
+
+		final CtTypeParameterReference generic = (CtTypeParameterReference) localTypes.get(0).getFormalTypeParameters().get(0);
+		assertNotNull(generic);
```
Comment (GerardPaligot) [inline src/test/java/spoon/test/type/TypeTest.java]: generic == T

File: src/main/java/spoon/reflect/declaration/CtTypeParameter.java [hunk 4]
```diff
@@ -30,22 +31,37 @@
...
 	 * @return
 	 */
+	@Deprecated
 	boolean removeBound(CtTypeReference<?> bounds);
+
+	/**
+	 * A type parameter can have an <code>extends</code> clause which declare
+	 * one ({@link CtTypeReference} or more ({@link CtIntersectionTypeReference} references.
+	 */
+	CtTypeReference<?> getSuperType();
```
Comment (GerardPaligot) [inline src/main/java/spoon/reflect/declaration/CtTypeParameter.java]: Rename getBoundingType

File: src/main/java/spoon/reflect/declaration/CtTypeParameter.java [hunk 5]
```diff
@@ -30,22 +31,37 @@
...
 	/**
 	 * @param bounds
 	 * @return
 	 */
+	@Deprecated
 	boolean removeBound(CtTypeReference<?> bounds);
+
+	/**
+	 * A type parameter can have an <code>extends</code> clause which declare
+	 * one ({@link CtTypeReference} or more ({@link CtIntersectionTypeReference} references.
```
Comment (GerardPaligot) [inline src/main/java/spoon/reflect/declaration/CtTypeParameter.java]: Add code example

File: src/main/java/spoon/reflect/reference/CtIntersectionTypeReference.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+package spoon.reflect.reference;
+
+import java.util.List;
+
+/**
+ * This interface defines a reference to an intersection type in generics or in casts.
+ */
+public interface CtIntersectionTypeReference<T> extends CtTypeReference<T> {
+	/**
+	 * Gets the bounds of the intersection type. Note that the first bound correspond to the current intersection type.
```
Comment (GerardPaligot) [inline src/main/java/spoon/reflect/reference/CtIntersectionTypeReference.java]: Add code example.

File: src/main/java/spoon/reflect/reference/CtIntersectionTypeReference.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	 */
+	List<CtTypeReference<?>> getBounds();
+
+	/**
+	 * Sets the bounds of the intersection type.
+	 */
+	<C extends CtIntersectionTypeReference> C setBounds(List<CtTypeReference<?>> bounds);
```
Comment (GerardPaligot) [inline src/main/java/spoon/reflect/reference/CtIntersectionTypeReference.java]: Remove

File: src/main/java/spoon/reflect/reference/CtIntersectionTypeReference.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.List;
+
+/**
+ * This interface defines a reference to an intersection type in generics or in casts.
+ */
+public interface CtIntersectionTypeReference<T> extends CtTypeReference<T> {
+	/**
+	 * Gets the bounds of the intersection type. Note that the first bound correspond to the current intersection type.
+	 */
+	List<CtTypeReference<?>> getBounds();
```
Comment (GerardPaligot) [inline src/main/java/spoon/reflect/reference/CtIntersectionTypeReference.java]: Defensive copy

File: src/test/java/spoon/test/intercession/IntercessionTest.java [hunk 9]
```diff
@@ -185,7 +186,7 @@ public boolean matches(CtMethod<?> element) {
 
 					if (!setter.getDeclaringType()
 							   .getSimpleName()
-							   .equals(((CtTypeParameterReference) ctTypeReference).getBounds().get(0).getSimpleName())) {
+							   .equals(((CtIntersectionTypeReference<?>) ((CtTypeParameterReference) ctTypeReference).getSuperType()).getBounds().get(0).getSimpleName())) {
```
Comment (GerardPaligot) [inline src/test/java/spoon/test/intercession/IntercessionTest.java]: use `asCtIntersectionTypeReference`.

File: src/test/java/spoon/test/type/TypeTest.java (hunk 3 above)
Comment (GerardPaligot) [inline src/test/java/spoon/test/type/TypeTest.java]: Done.

**Comment by GerardPaligot** [comment]
@monperrus PR ok.

File: src/main/java/spoon/reflect/reference/CtTypeParameterReference.java [hunk 10]
```diff
@@ -19,39 +19,63 @@
...
+	 * A type parameter can have an <code>extends</code> clause which declare
+	 * one ({@link CtTypeReference} or more ({@link CtIntersectionTypeReference} references.
+	 * <pre>
+	 *     // Extends with generics.
+	 *     T extends Interface1
+	 *     // Intersection type with generics.
+	 *     T extends Interface1 &amp; Interface2
+	 *     // Type in cast.
+	 *     (Runnable) () -&gt; System.out.println("");
+	 *     // Intersection type with casts.
```
Comment (monperrus) [inline src/main/java/spoon/reflect/reference/CtTypeParameterReference.java]: the cast is not handled there in CtTypeParameterReference?

**Comment by GerardPaligot** [comment]
@monperrus PR ok.

**Pull request merged**